[SPARK-26015][K8S] Set a default UID for Spark on K8S Images

rvesse opened the pull request:
## What changes were proposed in this pull request?

Adds USER directives to the Dockerfiles which is configurable via build argument (`spark_uid`) for easy customisation.  A `-u` flag is added to `bin/docker-image-tool.sh` to make it easy to customise this e.g.

```
> bin/docker-image-tool.sh -r rvesse -t uid -u 185 build
> bin/docker-image-tool.sh -r rvesse -t uid push
```

If no UID is explicitly specified it defaults to `185` - this is per @skonto's suggestion to align with the OpenShift standard reserved UID for Java apps ( 
https://lists.openshift.redhat.com/openshift-archives/users/2016-March/msg00283.html)

Notes:
- We have to make the `WORKDIR` writable by the root group or otherwise jobs will fail with `AccessDeniedException`

To Do:
- [x] Debug and resolve issue with client mode test
- [x] Consider whether to always propagate `SPARK_USER_NAME` to environment of driver and executor pods so `entrypoint.sh` can insert that into `/etc/passwd` entry
- [x] Rebase once PR #23013 is merged and update documentation accordingly

## How was this patch tested?

Built the Docker images with the new Dockerfiles that include the `USER` directives.  Ran the Spark on K8S integration tests against the new images.  All pass except client mode which I am currently debugging further.

Also manually dropped myself into the resulting container images via `docker run` and checked `id -u` output to see that UID is as expected.

Tried customising the UID from the default via the new `-u` argument to `docker-image-tool.sh` and again checked the resulting image for the correct runtime UID.

cc @felixcheung @skonto @vanzin

rvesse committed:
[SPARK-26015][K8S] Set a default UID for Spark on K8S Images

Adds USER directives to the Dockerfiles which is configurable via build
argument for easy customisation.  A -u flag is added to
bin/docker-image-tool.sh to make it easy to customise this.

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

rvesse commented:
For those with more knowledge of client mode here is the specific error seen in the integration tests:

```
Exception in thread "main" java.lang.IllegalArgumentException: basedir must be absolute: ?/.ivy2/local
  	at org.apache.ivy.util.Checks.checkAbsolute(Checks.java:48)
  	at org.apache.ivy.plugins.repository.file.FileRepository.setBaseDir(FileRepository.java:135)
  	at org.apache.ivy.plugins.repository.file.FileRepository.<init>(FileRepository.java:44)
  	at org.apache.spark.deploy.SparkSubmitUtils$.createRepoResolvers(SparkSubmit.scala:1063)
  	at org.apache.spark.deploy.SparkSubmitUtils$.buildIvySettings(SparkSubmit.scala:1149)
  	at org.apache.spark.deploy.DependencyUtils$.resolveMavenDependencies(DependencyUtils.scala:51)
  	at org.apache.spark.deploy.SparkSubmit.prepareSubmitEnvironment(SparkSubmit.scala:315)
  	at org.apache.spark.deploy.SparkSubmit.submit(SparkSubmit.scala:143)
  	at org.apache.spark.deploy.SparkSubmit.doSubmit(SparkSubmit.scala:86)
  	at org.apache.spark.deploy.SparkSubmit$$anon$2.doSubmit(SparkSubmit.scala:927)
  	at org.apache.spark.deploy.SparkSubmit$.main(SparkSubmit.scala:936)
  	at org.apache.spark.deploy.SparkSubmit.main(SparkSubmit.scala)
```

This looks to me like Spark/Ivy is discovering the users home directory incorrectly.  Have done some digging into the code paths but have not spotted what exactly is wrong yet.

felixcheung commented:
noted test issue. let's kick off test though

felixcheung commented:
ok to test

SparkQA commented:
**[Test build #98787 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/98787/testReport)** for PR 23017 at commit [`92e22f7`](https://github.com/apache/spark/commit/92e22f76bbe61035dcde54c5fb2990cb7099a98d).

ifilonenko commented:
> This looks to me like Spark/Ivy is discovering the users home directory incorrectly. Have done some digging into the code paths but have not spotted what exactly is wrong yet.

I stumbled upon this error as well, I solved this with the following config addition: 
```
--conf spark.driver.extraJavaOptions=-Divy.home=/root/.ivy2 \
```

vanzin commented on resource-managers/kubernetes/docker/src/main/dockerfiles/spark/entrypoint.sh:
```diff
@@ -30,6 +30,10 @@ set -e
 # If there is no passwd entry for the container UID, attempt to create one
 if [ -z "$uidentry" ] ; then
     if [ -w /etc/passwd ] ; then
+        # TODO Should we allow providing an environment variable to set the desired username?
```
I think it's ok to remove this TODO. The user name propagation in Hadoop should be fixed elsewhere (in fact I'm touching that code in another PR...).

But more interesting, by having a default UID, this code becomes a no-op, since `/etc/passwd` is not writable anymore.

And that's also probably the cause of the test failure you're seeing.

So maybe you need to call `useradd` or whatever is the appropriate command in this image to create that user when creating the image.

rvesse replied:
Not the case.  `USER` directives take effect at the point at which they occur in the `Dockerfile`, so since the directive is placed after the `ENTRYPOINT` directive the entry point script still runs as `root` allowing adding UID entries and any other privileged setup operations that are needed prior to the containers main process running and having its UID set appropriately

If you manually drop into the resulting container you can see that your UID is the specified non-privileged UID but that an `/etc/passwd` entry was successfully added for you by the entry point script

vanzin replied:
So aren't you basically saying that the entrypoint script is being run a root? Isn't that exactly what we do not want, since that will be running user code as root inside the image?

ifilonenko replied:
You can run the user code with a securityContext where you can specify runAs: {UID}, but without root you are unable to run useradd commands which would be crucial for said feature. Kubernetes defaults the security context to be root.
Also, is there a security problem with running as root in an isolated container?

vanzin replied:
@rvesse I tried your patch with a modified entrypoint just to print the UID. Here's what it prints:

```
AT START: 185
AT RUN POINT: 185, CMD: 
```

The patch:

```
--- a/resource-managers/kubernetes/docker/src/main/dockerfiles/spark/entrypoint.sh
+++ b/resource-managers/kubernetes/docker/src/main/dockerfiles/spark/entrypoint.sh
@@ -19,6 +19,8 @@
 # echo commands to the terminal output
 set -ex
 
+echo "AT START: $(id -u)"
+
 # Check whether there is a passwd entry for the container UID
 myuid=$(id -u)
 mygid=$(id -g)
@@ -91,6 +93,10 @@ if ! [ -z ${HADOOP_CONF_DIR+x} ]; then
   SPARK_CLASSPATH="$HADOOP_CONF_DIR:$SPARK_CLASSPATH";
 fi
 
+echo "AT RUN POINT: $(id -u), CMD: $SPARK_K8S_CMD"
+grep $(id -u) /etc/passwd
+exit 0
+
 case "$SPARK_K8S_CMD" in
   driver)
     CMD=(
```

So I think two things from that experiment:

- what I said is right, that the entrypoint is being run as the defined UID even if it's declared before the UID switch, and that code is a no-op if you defined the UID.
- docker seems to create the UID for you if it doesn't exist. either that or the image we're inheriting from already has that UID.

vanzin replied:
@ifilonenko 
> is there a security problem with running as root in an isolated container?

Other than it always being a horrible idea to run things as root, some deployments may disable containers that run as root from working. And if they do, and you mount some host directory into the container, the container has full privileges over that directory, which can be an issue.

In general, it's better to just avoid doing things as root since it raises more questions than it answers.

rvesse replied:
@vanzin Ok I was wrong here, you are correct that the UID set by the `USER` directive is in effect when the entry point runs.

This section still works because we don't see a UID so the user still belongs to the `root` group which has write access to the `/etc/passwd` file:

```
bash-4.4$ ls -l /etc/ | grep pass
-rw-rw-r--    1 root     root          1281 Nov 14 17:38 passwd
```

vanzin replied:
Actually my previous analysis is not completely correct. The entrypoint is indeed running as user 185. But:

```
bash-4.4$ ls -l /etc/passwd
-rw-rw-r--    1 root     root          1272 Nov 14 17:40 /etc/passwd
```

And:

```
bash-4.4$ id -u
185
bash-4.4$ id -g
0
```

So the group is still `root`, and user 185 can actually do a lot of things that only root should.

rvesse replied:
@vanzin Well I am mostly following advice via @skonto from https://docs.okd.io/latest/creating_images/guidelines.html#use-uid which says:

> For an image to support running as an arbitrary user, directories and files that may be written to by processes in the image should be owned by the root group and be read/writable by that group. Files to be executed should also have group execute permissions.
>
>Adding the following to your Dockerfile sets the directory and file permissions to allow users in the root group to access them in the built image:
> ...
>Because the container user is always a member of the root group, the container user can read and write these files. The root group does not have any special permissions (unlike the root user) so there are no security concerns with this arrangement. In addition, the processes running in the container must not listen on privileged ports (ports below 1024), since they are not running as a privileged user.

Which is also where this code snippet/pattern originates from the folks who originally developed the container images for Spark on K8S

vanzin replied:
> Because the container user is always a member of the root group

I guess that settles it. So what you have is fine. (Comment about the TODO not being needed still applies, though.)

SparkQA commented:
**[Test build #98787 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/98787/testReport)** for PR 23017 at commit [`92e22f7`](https://github.com/apache/spark/commit/92e22f76bbe61035dcde54c5fb2990cb7099a98d).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/98787/
Test FAILed.

rvesse committed:
[SPARK-26015][K8S] Fix broken client mode test

The client mode test was incorrectly overriding the entry point of the
image so didn't benefit from the logic that set up the /etc/passwd entry
for the container UID resulting in no home directory and a failed Ivy
setup as a result

rvesse committed:
[SPARK-26015][K8S] If SPARK_USER_NAME is present use it

If SPARK_USER_NAME is set for the pod then use it as part of the
/etc/passwd entry we create

SparkQA commented:
**[Test build #98872 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/98872/testReport)** for PR 23017 at commit [`a9e59b3`](https://github.com/apache/spark/commit/a9e59b3f9cbe9671da793e775aaa36d5df5ebcc6).

rvesse commented:
Resolved the issue with the client mode test.  The test itself was actually badly written in that it used the Spark images but overrode the entry point which avoided the logic that sets up the `/etc/passwd` entry for the container UID.  Without this running there is no home directory for the user so the Ivy setup fails as a result.  By changing the test to not override the entry point the test can run successfully.  Also made a couple of minor changes to this test to make it easier to debug.

rvesse commented:
> noted test issue. let's kick off test though

@felixcheung This is now resolved, please kick off a retest when you get chance

vanzin commented on resource-managers/kubernetes/docker/src/main/dockerfiles/spark/Dockerfile:
```diff
@@ -53,5 +54,9 @@ COPY data /opt/spark/data
 ENTRYPOINT [ "/opt/entrypoint.sh" ]
+
+# Specify the User that the actual main process will run as
+USER ${spark_uid}
```
nit: add line break

rvesse replied:
Done

vanzin commented on resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/ClientModeTestsSuite.scala:
```diff
@@ -17,13 +17,13 @@
-import scala.collection.JavaConverters._
 
 import org.apache.spark.deploy.k8s.integrationtest.KubernetesSuite.{k8sTestTag, INTERVAL, TIMEOUT}
+import org.scalatest.Tag
```
out of curiosity, why add the tag?

rvesse replied:
As I was just wanting to debug this specific failing test it was useful to run it on its own repeatedly as I investigated the issue.  I backed out other changes I made to enable this (like not cleaning up the tests driver pod) so I will back this out as well

vanzin replied:
BTW you don't need tags just to run specific tests.

e.g. in maven you can use `-Dsuites="*.SuiteName part of test name to test"

To do matching on the test name. Similar in sbt (`testOnly *.SuiteName -- -z part of test name to test`).

rvesse replied:
I used tags just because that's what the existing scripting did.  I thought there was probably a Maven property to do the same but using tags was fastest path for me.

vanzin reviewed (commented):
Do you want to remove WIP from the title now?

Not sure why integration tests didn't run. Will re-kick.

vanzin commented:
retest this please

SparkQA commented:
**[Test build #98882 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/98882/testReport)** for PR 23017 at commit [`a9e59b3`](https://github.com/apache/spark/commit/a9e59b3f9cbe9671da793e775aaa36d5df5ebcc6).

SparkQA commented:
**[Test build #98872 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/98872/testReport)** for PR 23017 at commit [`a9e59b3`](https://github.com/apache/spark/commit/a9e59b3f9cbe9671da793e775aaa36d5df5ebcc6).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/98872/
Test PASSed.

SparkQA commented:
**[Test build #98882 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/98882/testReport)** for PR 23017 at commit [`a9e59b3`](https://github.com/apache/spark/commit/a9e59b3f9cbe9671da793e775aaa36d5df5ebcc6).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/98882/
Test PASSed.

rvesse committed:
[SPARK-26015][K8S] Address PR Comments

- Add line breaks for clarity
- Remove extra test tag

rvesse committed:
[SPARK-26015][K8S] Fix line breaks

SparkQA commented:
**[Test build #98911 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/98911/testReport)** for PR 23017 at commit [`8f4fd19`](https://github.com/apache/spark/commit/8f4fd194e89a5062fc694cf917e1c2f744294495).

SparkQA commented:
**[Test build #98911 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/98911/testReport)** for PR 23017 at commit [`8f4fd19`](https://github.com/apache/spark/commit/8f4fd194e89a5062fc694cf917e1c2f744294495).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/98911/
Test PASSed.

vanzin commented:
LGTM but the integration tests *still* have not run.

vanzin commented:
ok to test

SparkQA commented:
**[Test build #98926 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/98926/testReport)** for PR 23017 at commit [`8f4fd19`](https://github.com/apache/spark/commit/8f4fd194e89a5062fc694cf917e1c2f744294495).

SparkQA commented:
**[Test build #98928 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/98928/testReport)** for PR 23017 at commit [`8f4fd19`](https://github.com/apache/spark/commit/8f4fd194e89a5062fc694cf917e1c2f744294495).

vanzin commented:
@shaneknapp  halp!

ifilonenko commented:
https://amplab.cs.berkeley.edu/jenkins/job/testing-k8s-prb-make-spark-distribution-unified/5086/ 
seems to be hanging on the distribution build (45min+)

SparkQA commented:
**[Test build #98932 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/98932/testReport)** for PR 23017 at commit [`8f4fd19`](https://github.com/apache/spark/commit/8f4fd194e89a5062fc694cf917e1c2f744294495).

shaneknapp commented:
@vanzin i'm really not sure what's going on w/this.  i noticed it happening on research-jenkins-worker-07 yesterday, so i rebooted the box and that seemed to fix it.

now it's back, and happening on ALL of the workers.  at least we know it's not something up w/the systems, but when it builds the dist.  

i'll try and build the dist manually on the workers and see what happens as well.

shaneknapp commented:
yep it gets to the same spot when i try and build manually, and fails:
```
[INFO] --- maven-source-plugin:3.0.1:test-jar-no-fork (create-source-jar) @ spark-mllib-local_2.12 ---
[INFO] Building jar: /home/eecs/sknapp/src/spark/mllib-local/target/spark-mllib-local_2.12-3.0.0-SNAPSHOT-test-sources.jar
```

vanzin commented:
manually on a different machine or just on the workers?

shaneknapp commented:
@vanzin @ifilonenko can you guys try and build the dist locally on your dev laptops?  here's a little wrapper script to make it easier:

you'll need to update your PATH to have some version of python3 installed (tho i don't actually think it's necessary), as well as JAVA_HOME...  might also need zinc in there as well.
```
#!/bin/bash

rm -f spark-*.tgz

export DATE=`date "+%Y%m%d"`
export REVISION=`git rev-parse --short HEAD`

export AMPLAB_JENKINS=1
export PATH="$PATH:/home/anaconda/envs/py3k/bin"

# Prepend JAVA_HOME/bin to fix issue where Zinc's embedded SBT incremental compiler seems to
# ignore our JAVA_HOME and use the system javac instead.
export PATH="$JAVA_HOME/bin:$PATH:/usr/local/bin"

# Generate random point for Zinc
export ZINC_PORT
ZINC_PORT=$(python -S -c "import random; print random.randrange(3030,4030)")

export SBT_OPTS="-Duser.home=$HOME -Dsbt.ivy.home=$HOME/.ivy2"
export SPARK_VERSIONS_SUITE_IVY_PATH="$HOME/.ivy2"

./dev/make-distribution.sh --name ${DATE}-${REVISION} --pip --tgz -DzincPort=${ZINC_PORT} \
     -Phadoop-2.7 -Pkubernetes -Pkinesis-asl -Phive -Phive-thriftserver
retcode=$?

exit $retcode
```

shaneknapp commented:
> manually on a different machine or just on the workers?

on a different machine (your laptop, something local, etc).

i'm trying on a couple of different workers and it's always hanging @ that exact step.  :\

nothing has been updated, system-wise, on ANY of the workers.  i think this might be a problem w/making the dist and am not looking forward to any git archaeology to find the broken change.  :(

shaneknapp commented:
btw i wiped all of my .ivy2 and .m2 dirs before building, just in case we're looking at a poisoned artifact.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/5089/
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/5086/
Test FAILed.

SparkQA commented:
**[Test build #98926 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/98926/testReport)** for PR 23017 at commit [`8f4fd19`](https://github.com/apache/spark/commit/8f4fd194e89a5062fc694cf917e1c2f744294495).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/98926/
Test PASSed.

SparkQA commented:
**[Test build #98928 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/98928/testReport)** for PR 23017 at commit [`8f4fd19`](https://github.com/apache/spark/commit/8f4fd194e89a5062fc694cf917e1c2f744294495).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/98928/
Test FAILed.

vanzin commented:
retest this please

SparkQA commented:
**[Test build #98943 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/98943/testReport)** for PR 23017 at commit [`8f4fd19`](https://github.com/apache/spark/commit/8f4fd194e89a5062fc694cf917e1c2f744294495).

SparkQA commented:
**[Test build #98932 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/98932/testReport)** for PR 23017 at commit [`8f4fd19`](https://github.com/apache/spark/commit/8f4fd194e89a5062fc694cf917e1c2f744294495).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/98932/
Test PASSed.

SparkQA commented:
**[Test build #98943 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/98943/testReport)** for PR 23017 at commit [`8f4fd19`](https://github.com/apache/spark/commit/8f4fd194e89a5062fc694cf917e1c2f744294495).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/98943/
Test FAILed.

rvesse committed:
[SPARK-26015][K8S] Updates docs for USER changes

rvesse committed:
[SPARK-26015][K8S] Fix incorrectly removed import

SparkQA commented:
**[Test build #99049 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/99049/testReport)** for PR 23017 at commit [`4fc40ca`](https://github.com/apache/spark/commit/4fc40ca43de6274f2efe33976977c1504e5db1e9).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/testing-k8s-prb-make-spark-distribution-unified/5177/

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/testing-k8s-prb-make-spark-distribution-unified/5177/

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/5177/
Test FAILed.

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/testing-k8s-prb-make-spark-distribution-unified/5178/

SparkQA commented:
Kubernetes integration test status success
URL: https://amplab.cs.berkeley.edu/jenkins/job/testing-k8s-prb-make-spark-distribution-unified/5178/

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/5178/
Test PASSed.

rvesse commented:
Have now added the doc updates to this so think this is ready for final review and merging

SparkQA commented:
**[Test build #99050 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/99050/testReport)** for PR 23017 at commit [`6dce8bb`](https://github.com/apache/spark/commit/6dce8bb7e471fdaa94da625a5ad2b457395a504c).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/99050/
Test PASSed.

SparkQA commented:
**[Test build #99049 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/99049/testReport)** for PR 23017 at commit [`4fc40ca`](https://github.com/apache/spark/commit/4fc40ca43de6274f2efe33976977c1504e5db1e9).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/99049/
Test PASSed.

vanzin commented on docs/running-on-kubernetes.md:
```diff
@@ -19,9 +19,9 @@ Please see [Spark Security](security.html) and the specific advice below before
 ## User Identity
 
-Images built from the project provided Dockerfiles do not contain any [`USER`](https://docs.docker.com/engine/reference/builder/#user) directives.  This means that the resulting images will be running the Spark processes as `root` inside the container.  On unsecured clusters this may provide an attack vector for privilege escalation and container breakout.  Therefore security conscious deployment ...
+Images built from the project provided Dockerfiles contain a default [`USER`](https://docs.docker.com/engine/reference/builder/#user) directive with a default UID of `185`.  This means that the resulting images will be running the Spark processes as this UID inside the container. Security conscious deployments should consider providing custom images with `USER` directives specifying their desired ...
```
Given the docs you quoted before, you can't override the container's GID, right?

rvesse replied:
Yes you can, the `USER` directive allows you to specify both a UID and a GID by separating them with a `:`.  However as noted here changing the GID is likely to have side effects that are difficult for Spark as a project to deal with hence the note about ensuring your UID belongs to the `root` group.  If end users need more complex setups then they are free to create their own custom images.

skonto replied:
I think using GUI 0 is safe. https://unix.stackexchange.com/questions/44077/what-does-it-mean-to-be-in-group-0

rvesse committed:
[SPARK-26015][K8S] Fix up changes to align with master changes

rvesse commented:
Rebased to catch up with master and adapt for @vanzin's improvements to Docker build context from PR #23019

SparkQA commented:
**[Test build #99445 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/99445/testReport)** for PR 23017 at commit [`8ab866b`](https://github.com/apache/spark/commit/8ab866b8f872a1d1ac0c43b374048c05893824c9).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/testing-k8s-prb-make-spark-distribution-unified/5518/

SparkQA commented:
Kubernetes integration test status success
URL: https://amplab.cs.berkeley.edu/jenkins/job/testing-k8s-prb-make-spark-distribution-unified/5518/

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/5518/
Test PASSed.

SparkQA commented:
**[Test build #99445 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/99445/testReport)** for PR 23017 at commit [`8ab866b`](https://github.com/apache/spark/commit/8ab866b8f872a1d1ac0c43b374048c05893824c9).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/99445/
Test PASSed.

vanzin commented:
Merging to master.
